ruma-common/src/push/condition.rs

```diff
@@ -42,13 +52,273 @@ pub enum PushCondition {
+        match self.find(pattern) {
+            Some(start) => {
+                let end = start + pattern.len();
+
+                // Look if the match has word boundaries.
+                let word_boundary_start = if !self.char_at(start).is_word_char() {
+                    true
+                } else {
+                    match self.find_prev_char(start) {
+                        Some(c) => !c.is_word_char(),
+                        None => true,
+                    }
+                };
+
+                if word_boundary_start {
+                    let word_boundary_end = end == self.len()
+                        || !self.find_prev_char(end).unwrap().is_word_char()
+                        || !self.char_at(end).is_word_char();
+
+                    if word_boundary_end {
+                        return true;
+                    }
+                }
+
+                // Find next word.
+                let non_word_str = &self[start..];
+                let non_word = match non_word_str.find(|c: char| !c.is_word_char()) {
+                    Some(pos) => pos,
+                    None => return false,
+                };
+
+                let word_str = &non_word_str[non_word..];
+                let word = match word_str.find(|c: char| c.is_word_char()) {
+                    Some(pos) => pos,
+                    None => return false,
+                };
+
+                (&word_str[word..]).matches_word(pattern)
+            }
+            None => false,
+        }
+    }
+}
+
+/// The flattened representation of a JSON object.
+#[derive(Clone, Debug)]
+pub struct FlattenedJson {
+    /// The internal map containing the flattened JSON as a pair path, value.
+    map: BTreeMap<String, String>,
+}
+
+impl FlattenedJson {
+    /// Create a `FlattenedJson` from `Raw`.
+    pub fn from_raw<T>(raw: &Raw<T>) -> Self
+    where
+        T: Serialize,
+    {
+        let mut s = Self { map: BTreeMap::new() };
+
+        s.flatten_value(to_json_value(raw).unwrap(), "".into());
+
+        s
+    }
+
+    /// Flatten and insert the `value` at `path`.
+    fn flatten_value(&mut self, value: JsonValue, path: String) {
+        match value {
+            JsonValue::Object(fields) => {
+                for (key, value) in fields {
+                    let path = if path.is_empty() { key } else { format!("{}.{}", path, key) };
+                    self.flatten_value(value, path);
+                }
+            }
+            JsonValue::String(s) => {
+                if self.map.insert(path.clone(), s).is_some() {
+                    warn!("Duplicate key in flattened JSON: {}", path);
+                };
+            }
+            JsonValue::Number(_) | JsonValue::Bool(_) => {
+                if self.map.insert(path.clone(), value.to_string()).is_some() {
+                    warn!("Duplicate key in flattened JSON: {}", path);
+                };
+            }
+            JsonValue::Array(_) | JsonValue::Null => {}
```

Comment:
I think if we allow matching on `true` or `false`, we should allow matching on `null` as well. Though now that I'm thinking about it, there won't be a distinction between `null` and `"null"` / `true` and `"true"` / `false` and `"false"`.

Anyways, I don't want to block this PR on that. I'll create an issue after this is merged.